### ISSUE TITLE
Enhancement: v6 `typeMatchesSpecifier` add ability to match via a pattern

### ISSUE THREAD

**Issue by RebeccaStevens** [typescript-eslint/typescript-eslint#6839]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Add ability to provide a regex expression of the type name instead of a string that has to match exactly. 
Additionally, allow for a glob pattern to be given as the "path" for the `FileSpecifier`.

### Fail

```typescript
// N/A
```

### Pass

```typescript
// Match all local types prefix with mutable
typeMatchesSpecifier(
  type,
  {
    from: "file",
    name: /^(M|m)utable.+/u,
  },
  program);

// Match all local types defined in a test or tests director.
typeMatchesSpecifier(
  type,
  {
    from: "file",
    name: /^.+/u,
    path: "**/tests?/**"
  },
  program);

// Match all build in "Readonly" types
typeMatchesSpecifier(
  type,
  {
    from: "lib",
    name: /^Readonly.+/u,
  },
  program);

// Match all types in all @foo/* packages
typeMatchesSpecifier(
  type,
  {
    from: "package",
    name: /^.+/u,
    package: /^@foo\/.+/u,
  },
  program);
```

### Additional Info

New specifier definitions will look something like this:

```ts
interface FileSpecifier {
  from: 'file';
  name: string | RegExp | (string | RegExp)[];
  path?: string;
}

interface LibSpecifier {
  from: 'lib';
  name: string | RegExp | (string | RegExp)[];
}

interface PackageSpecifier {
  from: 'package';
  name: string | RegExp | (string | RegExp)[];
  package: string | RegExp;
}
```

**Comment by RebeccaStevens** [comment]
@marekdedic

**Comment by bradzacher** [comment]
Globs are difficult because it means we have to add a dependency on a glob library and do fuzzy matches.

Could you please detail *why* you want these changes? What usecases are you using them for?

**Comment by RebeccaStevens** [comment]
globs are more of an extra I decided to include in this issue rather than making a new issue for them.
Regex support is the main part I'd like and would use.

**Comment by bradzacher** [comment]
> Regex support is the main part I'd like and would use.

Would you be able to describe what you want to use them for? What is the usecase you're trying to solve with the utility and how does regex make it easier?

**Comment by RebeccaStevens** [comment]
I guess the main thing is that I want to keep support for patterns when I add this feature to [`is-immutable-type`](https://github.com/RebeccaStevens/is-immutable-type) and to several rules in [`eslint-plugin-functional`](https://github.com/eslint-functional/eslint-plugin-functional).

One use case where this could be useful is to ignore all types from a certain package.

```ts
typeMatchesSpecifier(
  type,
  {
    from: "package",
    name: /^.+/u, // Match all
    package: "fp-ts",
  },
  program);
```

**Comment by RebeccaStevens** [comment]
Looking into this more, this might be a case where I'd want to implement my own version of `typeMatchesSpecifier` as there are quite a few extra features that [`is-immutable-type`](https://github.com/RebeccaStevens/is-immutable-type) currently supports. I don't think I'd be able to keep all these features without doing so (or at least now without this library's `typeMatchesSpecifier` becoming quite bloated).

Feel free to close this issue.

**Comment by marekdedic** [comment]
If I may add my 2¢: I think it would be worth it to make `typeMatchesSpecifier` more comprehensive and general (i.e. bloated :D) and add all the required features **iff** there is a clear need and usecase for it.

**Comment by RebeccaStevens** [comment]
> If I may add my 2¢: I think it would be worth it to make `typeMatchesSpecifier` more comprehensive and general (i.e. bloated :D) and add all the required features **iff** there is a clear need and usecase for it.

I've just make my first version of this if you want to see what I've changed and how bloaty it is: https://github.com/RebeccaStevens/is-immutable-type/blob/limit-types/src/utils.ts#L63-L111

I've made this whole additional file to deal with just calculating the type's name (and it's probably a little expensive)
https://github.com/RebeccaStevens/is-immutable-type/blob/limit-types/src/type-to-string.ts

**Comment by marekdedic** [comment]
Thanks, wouldn't some of that be useful here even without pattern matching? It seems to me that all the machinery around calculating type name isn't limited to that, or am I missing something? When originally implementing this, I used the simplest thing that seemed to work - seems like you understand this way more than me :D

**Comment by JoshuaKGoldberg** [comment]
I personally think it'd be reasonable to build this into typescript-eslint. Looking at the original use cases, being able to target DOM types for the `prefer-readonly-parameter-types` suggests allowing anything like `[/(.*)Element/, "Node"]` would be nice. But I've been pretty gung-ho about these changes - should hear from someone else in @typescript-eslint/triage-team too.

**Comment by JoshuaKGoldberg** [comment]
cc @bradzacher @Josh-Cena - any thoughts?

**Comment by bradzacher** [comment]
No skin off my nose if we want to add some more flexible ility to this!
